### PR TITLE
 MultigetGridDialect implementation

### PR DESCRIPTION
**Pull request by viktor44**
fix some tests

### CODE REVIEW

**Comment by schernolyas** [comment]
HI @Z-z-z-z !
I see that you migrated to Ignite 2.3.  Could you please upgrade Native Ignite Persistence ? I mean migration to DataStorageConfiguration because PersistentStoreConfiguration depricated.

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java
Comment (YevIgn) [inline ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java:136]: Maybe LinkedHashMap should be used here instead of HashMap? The order here seems to matter under PESSIMISTIC tx settings, where unordered getAll may easily cause deadlocks.

**Comment by viktor44** [comment]
> Could you please upgrade Native Ignite Persistence ? I mean migration to DataStorageConfiguration because PersistentStoreConfiguration depricated.

I don't think it is a good idea. In this case we will lose backward compatibility.

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java
Comment (viktor44) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java:136]: I'm not sure but may be you right. Fixed it

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java
Comment (schernolyas) [inline ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: What is reason of usinf PESSIMISTICS transaction?

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java
Comment (schernolyas) [inline ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java:32]: Test org.hibernate.ogm.backendtck.queries.QueriesWithAssociationsTest have fixed already. I will send it in next PR. Could you please remove the test?

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java
Comment (YevIgn) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java:136]: Thanks for the change! Also, here `idSet` will simply duplicate `ids` values as it will hold internally map and list, the same can be achieved using LinkedHashMap for `ids` and calling `entityCache.getAll( ids.values() )` thus requiring less intermediate structures allocation and method calls, as the LinkedHashMap is guaranteed to return values view that preserves insertion order.

Comment (viktor44) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java:136]: Your code isn't correct. getAll() requires Set, but values() is Collection.

Comment (YevIgn) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java:136]: You are right. I forgot that. Just in case - `getAll ( Collection )` is available on the IgniteInternalCache which is returned by IgniteEx#cachex method and I see that IgnitionEx and IgniteEx are already used by IgniteDatastoreProvider.

**Comment by viktor44** [comment]
If no objections can anybody merge this PR?

**Comment by DavideD** [comment]
@Z-z-z-z thanks, I will merge this before the end of the week

**Comment by schernolyas** [comment]
@Z-z-z-z Please change PESSIMISTICS transaction to OPTIMISTIC. 
Because PESSIMISTICS transactions hangs transactions

File: pom.xml
Comment (DavideD) [inline pom.xml]: Is this needed for this PR? You should keep all the changes in logical commits

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java
Comment (DavideD) [inline ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java:32]: Is this test related to the MultiGet implementation?

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/query/parsing/impl/IgnitePropertyHelper.java
Comment (DavideD) [inline ignite/src/main/java/org/hibernate/ogm/datastore/ignite/query/parsing/impl/IgnitePropertyHelper.java:90]: How come did you remove this part?

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java
Comment (DavideD) [inline ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: Is this needed for this PR? Or there is some other reason to change it?

**Review by DavideD — COMMENTED** [review]
The implementation is alright but there are some changes that don't seem necessary. We generally aim to keep changes in separate logical commits. This will make it easy for the reviewer to check for problems. I will wait for your answer to some of the questions before merging it

File: ignite/src/main/java/org/hibernate/ogm/datastore/ignite/query/parsing/impl/IgnitePropertyHelper.java
Comment (viktor44) [reply ignite/src/main/java/org/hibernate/ogm/datastore/ignite/query/parsing/impl/IgnitePropertyHelper.java:90]: this change was made to fix QueriesWithAssociationsTest.

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java
Comment (viktor44) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java:32]: No.

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java
Comment (YevIgn) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: This change seems to be here for assessment of the case, that no reordering of keys is done when they are passed to the Ignite in the actual "getTuples" method, as the "getAll" accepts set of keys, the set should maintain the same iteration order that the list of keys passed to the "getTuples" method, otherwise it might introduce another source of deadlock in case of PESSIMISTIC (upfront locking) concurrency level.

**Comment by YevIgn** [comment]
The PESSIMISTIC concurrency doesn't hangs tx by itself, rather it is done by varying order of entries access in different transactions, from the current state of Hibernate OGM Ignite it seems that this is the purely dependent on the user code.
Maybe there should just be a warning in case of PESSIMISTIC concurrency level is chosen, that the user should consider accounting for the cache entries access order and possibly adjust the parameters of deadlock detection - https://apacheignite.readme.io/docs/transactions#deadlock-detection - by also printing the given link to the warning?

File: pom.xml
Comment (viktor44) [reply pom.xml]: You right. Should I revert this change?

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java
Comment (viktor44) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: PESSIMISTIC had been in code from start of development this dialect and was changed 10 weeks ago. By mistake or misunderstanding, I believe. I have never had problems with tests due to PESSIMISTIC mode. I only corrected this mistake.

**Comment by schernolyas** [comment]
Hi @Z-z-z-z !

Tests works with OPTIMISTIC too. But I and @YevIgn  faced  tx hanging when we use PESSIMISTIC . It can be a problem of Apache Ignite ( tested versions are 1.9 and 2.1). Anyway  OPTIMISTIC  must be a default concurrency level because it most safe level.

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java
Comment (schernolyas) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: Tests works with OPTIMISTIC too. But I and @YevIgn faced tx hanging when we use PESSIMISTIC . It can be a problem of Apache Ignite ( tested versions are 1.9 and 2.1). Anyway OPTIMISTIC must be a default concurrency level because it most safe level.

File: pom.xml
Comment (schernolyas) [reply pom.xml]: Do you use 2.3 version on your project? You have to revert the change if you not need 2.3 version.

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java
Comment (schernolyas) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java:32]: Test org.hibernate.ogm.backendtck.queries.QueriesWithAssociationsTest have fixed already. I will send it in next PR. Could you please remove the test?

**Review by schernolyas — CHANGES_REQUESTED** [review]

File: pom.xml
Comment (DavideD) [reply pom.xml]: > You right. Should I revert this change?
 
Yes, please. Revert all the things that are not needed for the Multiget or, at least, extract them in separate commits so that I can check and apply and them separately.

Comment (viktor44) [reply pom.xml]: Done

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java
Comment (viktor44) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: What do you mean "safe level"?  OPTIMISTIC means no locking at all. I think it isn't serious to run tests with this transaction level

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java
Comment (viktor44) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/test/queries/QueriesWithAssociationsTest.java:32]: Why? I think it's correct. When you make your PR you will be able make to changes

**Comment by DavideD** [comment]
> Test org.hibernate.ogm.backendtck.queries.QueriesWithAssociationsTest have fixed already. I will send it in next PR. Could you please remove the test?

@schernolyas  Is there any difference between your solution and this one?

**Comment by DavideD** [comment]
> @Z-z-z-z Please change PESSIMISTICS transaction to OPTIMISTIC.
> Because PESSIMISTICS transactions hangs transactions

Mmmh... I would expect this to work in both scenarios. When did you experience the hanging of the transactions? Does it happen when you run the tests with maven?

**Comment by schernolyas** [comment]
HI @DavideD !
About "@schernolyas Is there any difference between your solution and this one?"
Yes. I fixed the test fully. And queries with join too.

About "Mmmh... I would expect this to work in both scenarios. When did you experience the hanging of the transactions? Does it happen when you run the tests with maven?"

It doesn't matter for tests. Both concurrency strategies works fine.  But ... I and @YevIgn  faced with problems with  PESSIMISTICS.  About experience ... Yes, we have. But we don't know why it occurs. As I wrote It is problem of Apache Ignite.

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java
Comment (YevIgn) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: Strictly speaking the issues, we faced with PESSIMISTIC setting, seem to be mostly related to the fact that in our case there was extensive use of one-to-many relationships, where two or more concurrent transactions accessed elements in different order. Or at least that is how I understood the problem back then.

**Comment by DavideD** [comment]
> It doesn't matter for tests. Both concurrency strategies works fine. But ... I and @YevIgn faced with problems with PESSIMISTICS. About experience ... Yes, we have. But we don't know why it occurs. As I wrote It is problem of Apache Ignite.

Mmmh... if that's the case I would prefer to have the tests on PESSIMISTIC so that we have some feedback if at some point they stop working for this scenario. Right now does not seem to be a problem.

**Comment by YevIgn** [comment]
Hello, Everyone.
From my point of view, there was no problem with Apache Ignite directly, rather the problem might be with Hibernate OGM Ignite or with the code that interacted with it on our side.
By the way, @DavideD, could you, please, advise - are there any best practice tips on how to handle the PESSIMISTIC locking strategy in general for many sided relationships? Are entity keys always passed in a some defined order to the underlying store? Can this order be altered/defined in user-code by using, e.g. @Order annotation?

**Comment by DavideD** [comment]
> From my point of view, there was no problem with Apache Ignite directly, rather the problem might be with Hibernate OGM Ignite or with the code that interacted with it on our side.

That sounds more likely.
It would be nice to have some tests if this is something that occurs. It will be also easier to answer follow-up questions.

> By the way, @DavideD, could you, please, advise - are there any best practice tips on how to handle the PESSIMISTIC locking strategy in general for many sided relationships?

Doesn't this depends more on the kind of application one is running? If one has operations hanging, it might be that he is too aggressive in acquiring the locks and everyone else has to wait until it's released.  But I might be wrong on this, so feel free to challenge it.
Also, Hibernate OGM/ORM relies on the underlying DB locking mechanism and there are ways to configure it, read the values form the dialect and then pass them.

> Are entity keys always passed in a some defined order to the underlying store? Can this order be altered/defined in user-code by using, e.g. @order annotation?

It depends which operations your are performing, I don't think there is a way to do it and I'm not sure is something the dialect should worry about. It seems the problem in this case is at an higher level than where Hibernate OGM/ORM is used.

File: ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java
Comment (DavideD) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: > OPTIMISTIC means no locking at all

Well, in general. It means that during an update or write the db will check that the previous data is not changed. It's a valid level and useful in many situations. We also have test in OGM to check for optimistic locking. I'm not sure how Apache Ignite handle it but maybe it can implement the `OptimisticLockingAwareGridDialect` interface.

Comment (DavideD) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: > Strictly speaking the issues, we faced with PESSIMISTIC setting, seem to be mostly related to the fact that in our case there was extensive use of one-to-many relationships, where two or more concurrent transactions accessed elements in different order. Or at least that is how I understood the problem back then.

Maybe too many fetch EAGER loading too many relationships and hitting too many entities?

Comment (YevIgn) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: >OPTIMISTIC means no locking at all

In case of Apache Ignite OPTIMISTIC means locks acquisition on prepare step. Version check is performed on SERIALIZABLE isolation level. Other combinations of concurrency/isolation levels are subject to deadlock detection mechanism - https://apacheignite.readme.io/docs/transactions#section-deadlock-detection

Comment (YevIgn) [reply ignite/src/test/java/org/hibernate/ogm/datastore/ignite/utils/IgniteTestConfigurationBuilder.java:43]: > Maybe too many fetch EAGER loading too many relationships and hitting too many entities?

It was certainly a deadlock, as the number of entities was fairly small - namely no more than a dozen, and it was actually quite simple to trace it to the different order of entries access.

**Comment by DavideD** [comment]
I've cherry picked the commits and change them a bit, therefore you won't find them on master.
I did the following:

* I've decided to leave the PESSIMISTIC approach for now, it seems that some had problem with it and it might be helpful if we can see the failure in the tests. Guys, if you can create a test case it would be great.

* Changed a bit the QueriesWithAssociationTest, it now extends the test in core

**Comment by DavideD** [comment]
Thanks @Z-z-z-z, feel free to send new PR if you see that I messed up somwhere. But Please, next time try to keep the changes in separate logical commits, you can have a look at the commits I created to see what I mean.

**Pull request closed without merging**
